resource/aws_cloudformation_stack_set_instance: IAM Role Trust Relationship Failures on Creation

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Description

Likely due to IAM eventual consistency, we regularly see these types of test failures:

```
=== CONT  TestAccAWSCloudFormationStackSetInstance_RetainStack
TestAccAWSCloudFormationStackSetInstance_RetainStack: resource_aws_cloudformation_stack_set_instance_test.go:251: Step 1/5 error: terraform failed: exit status 1
stderr:
Error: error waiting for CloudFormation StackSet Instance (tf-acc-test-7055553958409005411,*******,us-west-2) creation: Operation (terraform-20200915035156420300000004) Results:
Account (*******) Region (us-west-2) Status (FAILED) Status Reason: Account ******* should have 'tf-acc-test-7055553958409005411-Execution' role with trust relationship to Role 'tf-acc-test-7055553958409005411-Administration'.
--- FAIL: TestAccAWSCloudFormationStackSetInstance_RetainStack (45.93s)

=== CONT  TestAccAWSCloudFormationStackSetInstance_ParameterOverrides
TestAccAWSCloudFormationStackSetInstance_ParameterOverrides: resource_aws_cloudformation_stack_set_instance_test.go:188: Step 1/5 error: terraform failed: exit status 1
stderr:
Error: error waiting for CloudFormation StackSet Instance (tf-acc-test-3788714546865718265,*******,us-west-2) creation: Operation (terraform-20200915035155070000000004) Results:
Account (*******) Region (us-west-2) Status (FAILED) Status Reason: Account ******* should have 'tf-acc-test-3788714546865718265-Execution' role with trust relationship to Role 'tf-acc-test-3788714546865718265-Administration'.
--- FAIL: TestAccAWSCloudFormationStackSetInstance_ParameterOverrides (43.21s)

=== CONT  TestAccAWSCloudFormationStackSetInstance_disappears_StackSet
TestAccAWSCloudFormationStackSetInstance_disappears_StackSet: resource_aws_cloudformation_stack_set_instance_test.go:164: Step 1/1 error: terraform failed: exit status 1
stderr:
Error: error waiting for CloudFormation StackSet Instance (tf-acc-test-2511287766137543066,*******,us-west-2) creation: Operation (terraform-20200915035146166100000004) Results:
Account (*******) Region (us-west-2) Status (FAILED) Status Reason: Account ******* should have 'tf-acc-test-2511287766137543066-Execution' role with trust relationship to Role 'tf-acc-test-2511287766137543066-Administration'.
--- FAIL: TestAccAWSCloudFormationStackSetInstance_disappears_StackSet (39.39s)

=== CONT  TestAccAWSCloudFormationStackSetInstance_disappears
TestAccAWSCloudFormationStackSetInstance_disappears: resource_aws_cloudformation_stack_set_instance_test.go:140: Step 1/1 error: terraform failed: exit status 1
stderr:
Error: error waiting for CloudFormation StackSet Instance (tf-acc-test-8927522512012976761,*******,us-west-2) creation: Operation (terraform-20200915035142785300000004) Results:
Account (*******) Region (us-west-2) Status (FAILED) Status Reason: Account ******* should have 'tf-acc-test-8927522512012976761-Execution' role with trust relationship to Role 'tf-acc-test-8927522512012976761-Administration'.
--- FAIL: TestAccAWSCloudFormationStackSetInstance_disappears (43.98s)

=== CONT  TestAccAWSCloudFormationStackSetInstance_basic
TestAccAWSCloudFormationStackSetInstance_basic: resource_aws_cloudformation_stack_set_instance_test.go:106: Step 1/2 error: terraform failed: exit status 1
stderr:
Error: error waiting for CloudFormation StackSet Instance (tf-acc-test-7397349939751457548,*******,us-west-2) creation: Operation (terraform-20200915035141577400000004) Results:
Account (*******) Region (us-west-2) Status (FAILED) Status Reason: Account ******* should have 'tf-acc-test-7397349939751457548-Execution' role with trust relationship to Role 'tf-acc-test-7397349939751457548-Administration'.
--- FAIL: TestAccAWSCloudFormationStackSetInstance_basic (39.88s)
```

### Affected Resource(s)

* aws_cloudformation_stack_set_instance

--- a/aws/resource_aws_cloudformation_stack_set_instance.go
+++ b/aws/resource_aws_cloudformation_stack_set_instance.go
@@ -11,6 +11,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	iamwaiter "github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter"
 )
 
 func resourceAwsCloudFormationStackSetInstance() *schema.Resource {
@@ -95,16 +96,76 @@ func resourceAwsCloudFormationStackSetInstanceCreate(d *schema.ResourceData, met
 	}
 
 	log.Printf("[DEBUG] Creating CloudFormation StackSet Instance: %s", input)
-	output, err := conn.CreateStackInstances(input)
+	err := resource.Retry(iamwaiter.PropagationTimeout, func() *resource.RetryError {
+		output, err := conn.CreateStackInstances(input)
 
-	if err != nil {
-		return fmt.Errorf("error creating CloudFormation StackSet Instance: %s", err)
+		if err != nil {
+			return resource.NonRetryableError(fmt.Errorf("error creating CloudFormation StackSet Instance: %w", err))
+		}
+
+		d.SetId(fmt.Sprintf("%s,%s,%s", stackSetName, accountID, region))
+
+		err = waitForCloudFormationStackSetOperation(conn, stackSetName, aws.StringValue(output.OperationId), d.Timeout(schema.TimeoutCreate))
+
+		if err != nil {
+			// IAM eventual consistency
+			if strings.Contains(err.Error(), "AccountGate check failed") {
+				input.OperationId = aws.String(resource.UniqueId())
+				return resource.RetryableError(err)
+			}
+
+			// IAM eventual consistency
+			// User: XXX is not authorized to perform: cloudformation:CreateStack on resource: YYY
+			if strings.Contains(err.Error(), "is not authorized") {
+				input.OperationId = aws.String(resource.UniqueId())
+				return resource.RetryableError(err)
+			}
+
+			// IAM eventual consistency
+			// XXX role has insufficient YYY permissions
+			if strings.Contains(err.Error(), "role has insufficient") {
+				input.OperationId = aws.String(resource.UniqueId())
+				return resource.RetryableError(err)
+			}
+
+			// IAM eventual consistency
+			// Account XXX should have YYY role with trust relationship to Role ZZZ
+			if strings.Contains(err.Error(), "role with trust relationship") {
+				input.OperationId = aws.String(resource.UniqueId())
+				return resource.RetryableError(err)
+			}
+
+			// IAM eventual consistency
+			if strings.Contains(err.Error(), "The security token included in the request is invalid") {
+				input.OperationId = aws.String(resource.UniqueId())
+				return resource.RetryableError(err)
+			}
+
+			return resource.NonRetryableError(fmt.Errorf("error waiting for CloudFormation StackSet Instance (%s) creation: %w", d.Id(), err))
+		}
+
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		var output *cloudformation.CreateStackInstancesOutput
+		output, err = conn.CreateStackInstances(input)
+
+		if err != nil {
+			return fmt.Errorf("error creating CloudFormation StackSet Instance: %w", err)
+		}
+
+		d.SetId(fmt.Sprintf("%s,%s,%s", stackSetName, accountID, region))
+
+		err = waitForCloudFormationStackSetOperation(conn, stackSetName, aws.StringValue(output.OperationId), d.Timeout(schema.TimeoutCreate))
+
+		if err != nil {
+			return fmt.Errorf("error waiting for CloudFormation StackSet Instance (%s) creation: %w", d.Id(), err)
+		}
 	}
 
-	d.SetId(fmt.Sprintf("%s,%s,%s", stackSetName, accountID, region))
-
-	if err := waitForCloudFormationStackSetOperation(conn, stackSetName, aws.StringValue(output.OperationId), d.Timeout(schema.TimeoutCreate)); err != nil {
-		return fmt.Errorf("error waiting for CloudFormation StackSet Instance (%s) creation: %s", d.Id(), err)
+	if err != nil {
+		return err
 	}
 
 	return resourceAwsCloudFormationStackSetInstanceRead(d, meta)

